cic/clsp/filters/only_rekey.clsp

```diff
@@ -28,36 +28,32 @@
   (defun check_zero_coin (conditions curried_args_to_coin CURRIED_ARGS)
     (if conditions
         (c (f conditions)
-          (assert
-            (if (and  ; if this is an even CREATE_COIN
-                  (= CREATE_COIN (opcode (f conditions)))
-                  (not (logand (amount (f conditions)) 1))
+          (if (and  ; if this is an even CREATE_COIN
+                (= CREATE_COIN (opcode (f conditions)))
+                (not (logand (amount (f conditions)) 1))
+              )
+              (assert
+                ; it's a rekey coin
+                (not (amount (f conditions)))
+                (=
+                  (puzhash (f conditions))
+                  (puzzle_hash_of_curried_function (REKEY_MOD_HASH CURRIED_ARGS) (sha256tree curried_args_to_coin))
                 )
-                (assert
-                  ; it's a rekey coin
-                  (= 0 (amount (f conditions)))
-                  (=
-                    (puzhash (f conditions))
-                    (puzzle_hash_of_curried_function (REKEY_MOD_HASH CURRIED_ARGS) (sha256tree curried_args_to_coin))
-                  )
-                  ; with the correct timelock
-                  (not (> (ALLOWED_REKEY_TIMELOCK CURRIED_ARGS) (f (r (r curried_args_to_coin)))))
-                  ; then
-                  1
-                )
-                1
-            )
-            ; then
-            (check_zero_coin
-              (r conditions)
-              curried_args_to_coin
-              CURRIED_ARGS
-            )
+                ; with the correct timelock
+                (not (> (ALLOWED_REKEY_TIMELOCK CURRIED_ARGS) (f (r (r curried_args_to_coin)))))
+                ; then
+                (r conditions)  ; exit early if we find the even coin
```

Comment:
I don't like exiting early here.  In the current iteration, the prefarm inner puzzle will catch this, but this layer doesn't know about that and theoretically, this could let someone create one valid rekey drop coin first, and then a million invalid ones.